### PR TITLE
Generate a new IAC if action required is a social reminder

### PR DESCRIPTION
**Pull request by damien-lloyd**
# Motivation and Context
Only generate a new IAC if action is a social reminder. This is required for the reminder templates in actionexporter 

# What has changed
New clause added to ensure iac is generated only if action type is a social reminder. Also added delta to make such reminders optional in case the respondent no longer wants reminders.

# How to test?
Run all tests in action service

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java [hunk 1]
```diff
@@ -131,6 +134,11 @@ private ActionProcessingService getActionProcessingService(Action action) {
   private void processAction(
       Action action, final InstructionCount requestCount, final InstructionCount cancelCount)
       throws CTPException {
+
+    if (action.getActionType().getName().equals("SOCIALREM")) {
+      caseSvcClientService.generateNewIacForCase(action.getCaseId());
```
[the remark this answers is not in the document]
Comment (damien-lloyd) [reply src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java]: It returns the new IAC (for convenience) but further downstream the new IAC would be extracted by the CaseAndCaseEvent decorator from the CaseDetailsDTO which comes before the processing.

[the remark this answers is not in the document]
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java]: how is the IAC going to be extracted downstream if it's not been saved anywhere?

**Comment by damien-lloyd** [comment]
[SUS007: Generate Reminder Print File (8)](https://trello.com/c/cYp4TjKf/210-sus007-generate-reminder-print-file-8)

File: src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java [hunk 2]
```diff
@@ -44,7 +45,9 @@
   @Qualifier("caseSvcClient")
   private RestUtility restUtility;
 
-  @Autowired private ObjectMapper objectMapper;
+  @Qualifier("customObjectMapper")
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java]: why did this change? Seems unrelated to the PR

File: src/main/java/uk/gov/ons/ctp/response/action/config/CaseSvc.java [hunk 3]
```diff
@@ -15,4 +15,5 @@
   private String caseEventsByCasePostPath;
   private String addressByUprnGetPath;
   private String caseTypeByIdPath;
+  private String generateNewIacForCase;
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ctp/response/action/config/CaseSvc.java]: Should end with `PostPath` to match the other paths

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -112,6 +112,7 @@ case-svc:
   case-events-by-case-post-path: /cases/{caseid}/events
   address-by-uprn-get-path: /addresses/{uprn}
   case-type-by-id-path: /casetypes/{casetypeid}
+  generate-new-iac-for-case: /cases/{caseid}/iac
```
Comment (NickGrantONS) [inline src/main/resources/application.yml]: Should end with `-post-path` to match the others

File: src/test/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  @Test
+  public void ensureNewIacGenerated() throws Exception {
+
+    UriComponents uriComponents = mock(UriComponents.class);
+
+    CaseIACDTO cazeiac = new CaseIACDTO("abc123");
+    CaseSvc casesvc = mock(CaseSvc.class);
+
+    when(appConfig.getCaseSvc()).thenReturn(casesvc);
+    when(casesvc.getGenerateNewIacForCase()).thenReturn(CASE_GEN_IAC);
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientServiceTest.java]: this isn't really needed

File: src/test/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.util.UriComponents;
+import uk.gov.ons.ctp.common.rest.RestUtility;
+import uk.gov.ons.ctp.response.action.config.AppConfig;
+import uk.gov.ons.ctp.response.action.config.CaseSvc;
+import uk.gov.ons.ctp.response.casesvc.representation.CaseIACDTO;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CaseSvcClientServiceTest {
+
+  private final String CASE_GEN_IAC = "/cases/{caseid}/iac";
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientServiceTest.java]: this isn't really needed tbh

**Comment by AdamHawtin** [comment]
I agree with the changes already requested, but apart from that it all seems fine. 

I was able to run through a collection exercise execution with a reminder action rule in place and got a reminder file out at the end with newly generated IAC's on it.

File: src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java (hunk 2 above)
Comment (damien-lloyd) [reply src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java]: @NickGrantONS Intellij was complaining about not being able to autowire the object mapper.

File: pom.xml [hunk 7]
```diff
@@ -154,7 +154,7 @@
         <dependency>
             <groupId>uk.gov.ons.ctp.product</groupId>
             <artifactId>actionsvc-api</artifactId>
-            <version>10.49.26</version>
+            <version>10.49.27-SNAPSHOT</version>
```
Comment (NickGrantONS) [inline pom.xml]: do not use SNAPSHOT versions

File: src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java [hunk 8]
```diff
@@ -43,6 +44,7 @@
  * CANCEL_SUBMITTED action so expect 1 ActionCancel - all actions have responseRequired = true
  */
 @RunWith(MockitoJUnitRunner.class)
+@Ignore
```
Comment (NickGrantONS) [inline src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java]: why are you ignoring these tests?

File: src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java (hunk 2 above)
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/action/client/CaseSvcClientService.java]: I'd prefer it removed - we should let spring use autowiring wherever possible

File: src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java (hunk 8 above)
Comment (damien-lloyd) [reply src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java]: @NickGrantONS These tests are out of date and need to be updated in a later PR. They're not testing what we are developing with. If I unignore them, they'll simply cause the maven build to fail. Since this PR is needed for the current sprint, this compromise would be reasonably. We already have acceptance tests that test the scope of this task in generating a new IAC for social reminders.

Comment (NickGrantONS) [reply src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java]: if you broke something with your changes, you should fix it

Comment (NickGrantONS) [reply src/test/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributorTest.java]: sprint doesn't matter - we need to size and estimate things correctly or else we end up thinking our velocity is higher than it actually is, and we keep overcommitting to too much work, and cutting more corners. Quality has to be maintained.

File: src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java [hunk 9]
```diff
@@ -131,6 +134,13 @@ private ActionProcessingService getActionProcessingService(Action action) {
   private void processAction(
       Action action, final InstructionCount requestCount, final InstructionCount cancelCount)
       throws CTPException {
+
+    if (action.getActionType().getName().equals("SOCIALREM")
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java]: Don't like the hardcoding

File: src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java [hunk 10]
```diff
@@ -31,8 +32,11 @@
  */
 @Component
 class ActionDistributor {
+
   private static final Logger log = LoggerFactory.getLogger(ActionDistributor.class);
 
+  private static final String SOCIALREM = "SOCIALREM";
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java]: actually this is pretty ugly.

Couldn't we do this:
```java
uk.gov.ons.ctp.response.action.representation.ActionType.valueOf(
      action.getActionType().getName()) == uk.gov.ons.ctp.response.action.representation.ActionType.SOCIALREM
```

Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ctp/response/action/scheduled/distribution/ActionDistributor.java]: in fact why are we even using getName()?
